File: Backend/scripts/fetch_myntu_links.py
#!/usr/bin/env python3
"""
fetch_myntu_links.py
--------------------
從 myNTU 官方 API (mainHandler.ashx) 抓取所有分類連結（中英文）。
輸出格式替換 Frontend/public/links.json，同時供 import_links 匯入 DB。

使用方式（從 Backend 目錄執行）:
    python scripts/fetch_myntu_links.py
或透過 sync_data.sh:
    ./sync_data.sh links
"""

import json
import sys
import time
from pathlib import Path

import requests
import urllib3

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

API_URL = "https://my.ntu.edu.tw/mainHandler.ashx"
HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/120.0.0.0 Safari/537.36"
    )
}

# 所有分類（中文名稱 → 英文名稱），順序即前端顯示順序
CATEGORIES = [
    {"id": "students",  "zh": "學生專區",  "en": "Students",   "icon": "🎓"},
    {"id": "courses",   "zh": "課程學習",  "en": "Courses",    "icon": "📚"},
    {"id": "faculty",   "zh": "教職申辦",  "en": "Faculty & Staff", "icon": "👔"},
    {"id": "teaching",  "zh": "教學",      "en": "Teaching",   "icon": "🏫"},
    {"id": "library",   "zh": "圖書研究",  "en": "Research",   "icon": "🔬"},
    {"id": "finance",   "zh": "帳務財物",  "en": "Accounts",   "icon": "💰"},
    {"id": "venue",     "zh": "場館交通",  "en": "Facilities", "icon": "🏟️"},
    {"id": "campus",    "zh": "校園資源",  "en": "Resources",  "icon": "🏫"},
    {"id": "news",      "zh": "消息公告",  "en": "Bulletin",   "icon": "📢"},
    {"id": "feedback",  "zh": "意見交流",  "en": "Opinions",   "icon": "💬"},
]

DEFAULT_ICON_URL = "https://my.ntu.edu.tw/nasattach/0.png"

def check_image_exists(url: str, session: requests.Session) -> bool:
    try:
        response = session.head(url, verify=False, timeout=3)
        return response.status_code == 200
    except requests.exceptions.RequestException:
        return False


def create_session(lang_param: str) -> requests.Session:
    """建立帶語言 cookie 的 Session。"""
    session = requests.Session()
    session.headers.update(HEADERS)
    init_url = f"https://my.ntu.edu.tw/Default.aspx?lang={lang_param}"
    try:
        session.get(init_url, verify=False, timeout=15)
    except Exception as e:
        print(f"  ⚠ 初始化 session 失敗 ({lang_param}): {e}", file=sys.stderr)
    return session


def fetch_category(session: requests.Session, search_name: str) -> list[dict]:
    """呼叫 API 取得單一分類的所有連結。"""
    try:
        resp = session.post(
            API_URL,
            data={"type": "more", "searchName": search_name},
            verify=False,
            timeout=15,
        )
        resp.raise_for_status()
        return resp.json()
    except Exception as e:
        print(f"  ⚠ 取得 '{search_name}' 失敗: {e}", file=sys.stderr)
        return []


def build_icon_url(sys_id: str | int) -> str:
    return f"https://my.ntu.edu.tw/nasattach/{sys_id}.png"


def main() -> None:
    script_dir = Path(__file__).resolve().parent
    project_root = script_dir.parent.parent
    output_path = project_root / "Frontend" / "public" / "links.json"

    print("=" * 60)
    print("myNTU Links Scraper")
    print("=" * 60)

    # ── 階段一：抓中文資料 ──────────────────────────────────
    print("\n[1/2] 抓取中文資料...")
    session_zh = create_session("cht")
    raw: dict[str, dict] = {}  # sys_id → record

    for cat in CATEGORIES:
        items = fetch_category(session_zh, cat["zh"])
        for item in items:
            sys_id = str(item.get("sys_id", ""))
            if not sys_id or sys_id in raw:
                continue

            guessed_icon_url = build_icon_url(sys_id)
            if check_image_exists(guessed_icon_url, session_zh):
                final_icon_url = guessed_icon_url
            else:
                final_icon_url = DEFAULT_ICON_URL

            raw[sys_id] = {
                "sys_id":      sys_id,
                "category_id": cat["id"],
                "name_zh":     (item.get("sys_cname") or "").strip(),
                "name_en":     "",
                "link_zh":     (item.get("href") or "").strip(),
                "link_en":     "",
                "icon_url":    final_icon_url,
            }
        print(f"  {cat['zh']}: {len(items)} 筆")
        time.sleep(0.15)

    print(f"  → 共 {len(raw)} 筆唯一連結")

    # ── 階段二：抓英文資料並合併 ────────────────────────────
    print("\n[2/2] 抓取英文資料並合併...")
    session_en = create_session("eng")

    for cat in CATEGORIES:
        items = fetch_category(session_en, cat["en"])
        for item in items:
            sys_id = str(item.get("sys_id", ""))
            if sys_id in raw:
                raw[sys_id]["name_en"] = (item.get("sys_cname") or "").strip()
                raw[sys_id]["link_en"] = (item.get("href") or "").strip()
        print(f"  {cat['en']}: {len(items)} 筆")
        time.sleep(0.15)

    # ── 組裝輸出格式 ────────────────────────────────────────
    # 依分類分組
    grouped: dict[str, list[dict]] = {cat["id"]: [] for cat in CATEGORIES}
    for record in raw.values():
        cat_id = record["category_id"]
        if cat_id not in grouped:
            continue
        grouped[cat_id].append({
            "label":    record["name_zh"],
            "label_en": record["name_en"],   # 空字串代表英文版不顯示
            "url":      record["link_zh"],
            "url_en":   record["link_en"],   # 英文版專屬連結（可與中文相同）
            "icon":     record["icon_url"],
        })

    output = []
    for cat in CATEGORIES:
        output.append({
            "id":       cat["id"],
            "icon":     cat["icon"],
            "label":    cat["zh"],
            "label_en": cat["en"],
            "links":    grouped[cat["id"]],
        })

    # ── 寫出 JSON ───────────────────────────────────────────
    with output_path.open("w", encoding="utf-8") as f:
        json.dump(output, f, ensure_ascii=False, indent=2)

    total_links = sum(len(cat["links"]) for cat in output)
    print(f"\n✅ 完成！共 {total_links} 筆連結寫入 {output_path}")
    print("\n分類統計：")
    for cat in output:
        n = len(cat["links"])
        en_count = sum(1 for l in cat["links"] if l["label_en"])
        print(f"  {cat['icon']} {cat['label']} ({cat['id']}): {n} 筆，英文名稱 {en_count}/{n}")


if __name__ == "__main__":
    main()

File: Backend/scripts/fetch_myntu_links.ts
import {
  LINK_CATEGORIES,
  MYNTU_API_URL,
  upsertLinksWithEmbeddings,
} from "./common";

// ─── Config ───────────────────────────────────────────────────────────────────

const MYNTU_BASE_URL = "https://my.ntu.edu.tw";
const DEFAULT_ICON_URL = `${MYNTU_BASE_URL}/nasattach/0.png`;

const USER_AGENT =
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";

const OPENROUTER_API_KEY = process.env.OPENROUTER_API_KEY ?? "";
const EMBED_MODEL = "nvidia/llama-nemotron-embed-vl-1b-v2:free";
export const EMBED_DIMS = 2048;

// ─── Types ────────────────────────────────────────────────────────────────────

/** A link item that already carries its embedding vectors — passed to upsertLinksWithEmbeddings */
export interface LinkItemWithEmbeddings {
  label: string;
  label_en: string | null;
  url: string;
  url_en: string;
  icon: string;
  keywords: string;
  embeddings: number[][];
}

export interface LinkCategoryWithEmbeddings {
  id: string;       // slug
  icon: string;
  label: string;
  label_en: string | null;
  links: LinkItemWithEmbeddings[];
}

// ─── Session (cookie warm-up) ─────────────────────────────────────────────────
//
// The NTU API (mainHandler.ashx) determines response language from a session
// cookie set by Default.aspx, NOT from the searchName or any request header.
// Without this warm-up, every request returns Chinese data regardless of lang.
// This mirrors Python's create_session() which does a real GET to Default.aspx.

interface NtuSession {
  cookie: string;
  lang: string;
}

async function createSession(langParam: "cht" | "eng"): Promise<NtuSession> {
  const initUrl = `${MYNTU_BASE_URL}/Default.aspx?lang=${langParam}`;
  let cookie = "";

  try {
    const res = await fetch(initUrl, {
      method: "GET",
      headers: { "User-Agent": USER_AGENT },
      redirect: "follow",
    });

    // Extract all Set-Cookie headers and forward them on subsequent requests
    const setCookie = res.headers.getSetCookie?.() ?? [];
    cookie = setCookie
      .map((c) => c.split(";")[0]) // keep only name=value, strip path/domain/expires
      .join("; ");

    if (!cookie) {
      console.warn(`  ⚠ No cookies received for lang=${langParam} — English data may be empty`);
    }
  } catch (err) {
    console.warn(`  ⚠ Session warm-up failed for lang=${langParam}: ${err}`);
  }

  return { cookie, lang: langParam };
}

// ─── myNTU API ────────────────────────────────────────────────────────────────

async function fetchCategory(
  session: NtuSession,
  searchName: string
): Promise<Array<Record<string, unknown>>> {
  try {
    const res = await fetch(MYNTU_API_URL, {
      method: "POST",
      headers: {
        "User-Agent": USER_AGENT,
        "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
        // Forward the session cookie — this is the key difference vs the old TS version
        ...(session.cookie ? { Cookie: session.cookie } : {}),
      },
      body: new URLSearchParams({ type: "more", searchName }).toString(),
    });

    if (!res.ok) throw new Error(`HTTP ${res.status} ${res.statusText}`);

    return (await res.json()) as Array<Record<string, unknown>>;
  } catch (err) {
    console.warn(`  ⚠ Failed to fetch "${searchName}": ${err}`);
    return [];
  }
}

function buildIconUrl(sysId: string): string {
  return `${MYNTU_BASE_URL}/nasattach/${sysId}.png`;
}

async function checkImageExists(url: string): Promise<boolean> {
  try {
    const res = await fetch(url, { method: "HEAD" });
    return res.ok;
  } catch {
    return false;
  }
}

function normalizeText(value: unknown): string {
  if (value === null || value === undefined) return "";
  return String(value).trim();
}

// ─── Embedding ────────────────────────────────────────────────────────────────

async function embedTexts(
  texts: string[],
  inputType: "passage" | "query" = "passage"
): Promise<number[][]> {
  if (!OPENROUTER_API_KEY) {
    throw new Error(
      "Missing OPENROUTER_API_KEY — add it to your .env\n" +
        "Get a free key at https://openrouter.ai/keys"
    );
  }

  const BATCH_SIZE = 32;
  const results: number[][] = [];

  for (let i = 0; i < texts.length; i += BATCH_SIZE) {
    const batch = texts.slice(i, i + BATCH_SIZE);

    const res = await fetch("https://openrouter.ai/api/v1/embeddings", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${OPENROUTER_API_KEY}`,
      },
      body: JSON.stringify({
        model: EMBED_MODEL,
        input: batch,
        input_type: inputType,
        modality: batch.map(() => "text"),
      }),
    });

    if (!res.ok) {
      const body = await res.text();
      throw new Error(`OpenRouter embedding failed [${res.status}]: ${body}`);
    }

    const json = (await res.json()) as {
      data: Array<{ index: number; embedding: number[] }>;
    };

    const sorted = [...json.data].sort((a, b) => a.index - b.index);
    results.push(...sorted.map((d) => d.embedding));

    if (i + BATCH_SIZE < texts.length) {
      await new Promise((r) => setTimeout(r, 200));
    }
  }

  return results;
}

// ─── Main ─────────────────────────────────────────────────────────────────────

export async function main() {
  console.log("============================================================");
  console.log("myNTU Links Scraper + Nemotron Embedder");
  console.log(`Embedding model : ${EMBED_MODEL}`);
  console.log(`Embedding dims  : ${EMBED_DIMS}`);
  console.log("============================================================");

  if (!OPENROUTER_API_KEY) {
    console.error("\n❌ OPENROUTER_API_KEY is not set. Add it to your .env file.\n");
    process.exit(1);
  }

  // ── Phase 1: scrape Chinese data ──────────────────────────────────────────
  console.log("\n[1/3] Initialising Chinese session + fetching data...");
  const sessionZh = await createSession("cht");

  const raw = new Map<
    string,
    {
      category_id: string;
      label: string;
      label_en: string;
      url: string;
      url_en: string;
      icon: string;
    }
  >();

  for (const category of LINK_CATEGORIES) {
    const items = await fetchCategory(sessionZh, category.zh);
    for (const item of items) {
      const sysId = normalizeText(item.sys_id);
      if (!sysId || raw.has(sysId)) continue;

      // Mirror Python: check if icon image actually exists, fall back to default
      const iconUrl = buildIconUrl(sysId);
      const iconExists = await checkImageExists(iconUrl);

      raw.set(sysId, {
        category_id: category.id,
        label: normalizeText(item.sys_cname),
        label_en: "",
        url: normalizeText(item.href),
        url_en: "",
        icon: iconExists ? iconUrl : DEFAULT_ICON_URL,
      });
    }
    console.log(`  ${category.zh}: ${items.length} items`);
    await new Promise((r) => setTimeout(r, 150)); // mirrors Python's time.sleep(0.15)
  }

  console.log(`  → ${raw.size} unique links`);

  // ── Phase 2: merge English data ───────────────────────────────────────────
  // KEY FIX: warm up a real English session first so the lang cookie is set,
  // then pass it to every fetchCategory call for English.
  // Mirrors Python's: session_en = create_session("eng")
  console.log("\n[2/3] Initialising English session + merging data...");
  const sessionEn = await createSession("eng");

  for (const category of LINK_CATEGORIES) {
    const items = await fetchCategory(sessionEn, category.en);
    for (const item of items) {
      const sysId = normalizeText(item.sys_id);
      if (!raw.has(sysId)) continue;
      const current = raw.get(sysId)!;
      current.label_en = normalizeText(item.sys_cname);
      current.url_en = normalizeText(item.href);
    }
    console.log(`  ${category.en}: ${items.length} items`);
    await new Promise((r) => setTimeout(r, 150));
  }

  // ── Phase 3: generate embeddings ──────────────────────────────────────────
  console.log(`\n[3/3] Generating embeddings via OpenRouter...`);

  // Batch all texts together for efficiency — track count per item to re-split later
  const allTexts: string[] = [];
  const itemMeta: Array<{ sysId: string; count: number }> = [];

  for (const [sysId, record] of raw) {
    // Embed zh + en label + category labels as extra context
    // Matches the current upsertLinks logic: [label, label_en, cat.label, cat.label_en]
    const cat = LINK_CATEGORIES.find((c) => c.id === record.category_id);
    const texts = [
      record.label,
      record.label_en,
      cat?.zh ?? "",
      cat?.en ?? "",
    ].filter(Boolean) as string[];

    allTexts.push(...texts);
    itemMeta.push({ sysId, count: texts.length });
  }

  console.log(`  ${allTexts.length} texts to embed across ${raw.size} items`);
  const allVectors = await embedTexts(allTexts, "passage");

  // Re-associate vectors back to their items
  const embeddingMap = new Map<string, number[][]>();
  let offset = 0;
  for (const { sysId, count } of itemMeta) {
    embeddingMap.set(sysId, allVectors.slice(offset, offset + count));
    offset += count;
  }
  console.log(`  ✓ ${embeddingMap.size} items embedded`);

  // ── Assemble output ───────────────────────────────────────────────────────
  const grouped = new Map<string, LinkCategoryWithEmbeddings>();
  for (const cat of LINK_CATEGORIES) {
    grouped.set(cat.id, {
      id: cat.id,
      icon: cat.icon,
      label: cat.zh,
      label_en: cat.en,
      links: [],
    });
  }

  for (const [sysId, record] of raw) {
    const category = grouped.get(record.category_id);
    if (!category) continue;

    const keywords = [record.label, record.label_en]
      .filter(Boolean)
      .join(" ");

    category.links.push({
      label: record.label,
      label_en: record.label_en || null,
      url: record.url,
      url_en: record.url_en,
      icon: record.icon,
      keywords,
      embeddings: embeddingMap.get(sysId) ?? [],
    });
  }

  const categories = Array.from(grouped.values());
  const totalLinks = categories.reduce((s, c) => s + c.links.length, 0);

  // Print summary mirroring Python's output
  console.log(`\n  ${totalLinks} links across ${categories.length} categories:`);
  for (const cat of categories) {
    const enCount = cat.links.filter((l) => l.label_en).length;
    console.log(
      `  ${cat.icon} ${cat.label} (${cat.id}): ${cat.links.length} links, EN labels ${enCount}/${cat.links.length}`
    );
  }

  // ── Upsert into DB ────────────────────────────────────────────────────────
  const result = await upsertLinksWithEmbeddings(categories);
  console.log(
    `\n✅ Done! ${result.links} links across ${result.categories} categories synced.`
  );
}

if (import.meta.main) {
  main().catch((err) => {
    console.error("\n❌ Fatal:", err);
    process.exit(1);
  });
}